clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
             int totalBytes = 0;
             int totalCount = 0;
             FetchResponse response = new FetchResponse(resp.responseBody());
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
+                TopicPartition tp = entry.getKey();
+                FetchResponse.PartitionData partition = entry.getValue();
+                long fetchOffset = request.fetchData().get(tp).offset;
+                if (subscriptions.assignedPartitions().contains(tp) && partition.errorCode == Errors.OFFSET_OUT_OF_RANGE.code())
+                    offsetOutOfRangePartitions.put(tp, fetchOffset);
+            }
+            if (!offsetOutOfRangePartitions.isEmpty())
+                this.exceptionFromLastFetch = new OffsetOutOfRangeException(offsetOutOfRangePartitions);
```

Comment:
I think both work well. But I slightly prefer the current approach. If the only reason we want to store the map is because we will create the exception afterwards, why not do it in the first place? In the future we might want to store other exception as well. What do you think?